### PR TITLE
feat: add permissions & voting power per roles per features

### PR DESCRIPTION
**Pull request by MikaelVallenet**
# Add permissions & voting power per roles per resources

## PR Description

Hello, This PR have for goal to allow people to assign resources to roles they create for their DAOs
I do this through a modal where user can check what resource they want for the role they are creating

image

Then when a user with the roles votes it looks like this for now (For now user can just assign 999 voting power to roles, but we have to think of a way to handle this as a user input when creating the roles)

It would be nice to have a way to display this in a more intuitive / explicite way, need to be discussed

image

## What have be done

<details><summary>DoD checklist...</summary>

- [x] Modify the Interface ``IVotingModule`` to accept a feature when retrieving voting power
- [x] Create a new implementation of this interface called ``RolesVotingModule``
- [x] New implementation works almost the same as the classic one but communicate with role manager to get informations about the power of the roles
- [x] The voting power of the roles is handle in the wrapper of role manager for dao
- [x] Add NewRoleJSON Wrapper method to expect a json array of ``{resource, power}``
- [x] Change UI to make a modal IAM like that when creating a role you can hit some checkbox to give power on some features
- [x] Add a new template of DAO code generation using the roles voting group
- [x] Use this new template in roles based TORG
- [x] Now, a user with permissions on a feature could be able to "bypass" proposals (still need to create the proposal, but the user could execute right after that)
- [x] Add New Gno Test
- [x] Add Checkboxes to allow creator to choose what resources to assign to roles
</details>

### Add a user input to set the power of a role (numeric) - assign to me @MikaelVallenet 

image

## Change the Roles Settings Page in the creation form flow

image

I would like a table to see the roles with a details modal with all roles when we click on it

### Dropdown to allow to assign user with roles created 
Use a select input in `RolesSettingsSection`, plugged to the previously created roles

image

### Add a Roles Page that display each roles settings based on UI of an IAM/excel table

image

### Use the color choose to display the roles
image

Change the text input by a dropdown multi selector

### Color Picker to choose the role color
Create a color picker and use it in `MultisigMembers`

image

### Enhance Roles Review

Display resources linked to roles in a better way

### CODE REVIEW

File: gno/p/role_manager/role_manager.gno [hunk 1]
```diff
@@ -163,6 +163,30 @@ func (rm *RoleManager) HasRole(user std.Address, roleName string) bool {
 	return userRoles.Has(roleName)
 }
 
+func (rm *RoleManager) IsRoleExist(roleName string) bool {
```
Comment (n0izn0iz) [inline gno/p/role_manager/role_manager.gno:166]:
```suggestion
func (rm *RoleManager) RoleExists(roleName string) bool {
```

**Review by n0izn0iz — CHANGES_REQUESTED** [review]

File: gno/p/role_manager/role_manager.gno (hunk 1 above)
Comment (MikaelVallenet) [reply gno/p/role_manager/role_manager.gno]: [d8ae713](https://github.com/TERITORI/teritori-dapp/pull/1443/commits/d8ae71312deb89d35043dc4451eab17eca0a735e)

File: packages/utils/types/organizations.ts [hunk 2]
```diff
@@ -42,7 +42,7 @@ export type MembershipMemberSettingFormType = {
 // ROLES BASED ORGANIZATION FORM TYPES
 
 export type RolesSettingFormType = {
-  roles: { name: string; color: string }[];
+  roles: { name: string; color: string, features: string[] }[];
```
Comment (n0izn0iz) [inline packages/utils/types/organizations.ts:45]: can we rename "feature" to "ressource" (or something else) in the code? we already have the concept of `NetworkFeature` and I'm afraid this naming will cause confusion

File: gno/p/dao_interfaces/modules.gno [hunk 3]
```diff
@@ -44,7 +44,7 @@ type IRolesModule interface {
 	Render(path string) string
 	GetMemberRoles(address std.Address) []string
 	HasRole(address std.Address, role string) bool
-	NewRole(roleName string)
+	NewRole(roleName string, permissionsJSON string)
```
Comment (n0izn0iz) [inline gno/p/dao_interfaces/modules.gno:47]: we shouldn't take json in this func, if we need to pass json through the interface, create a `NewRoleJSON` func instead

JSON un/marshaling should be only at the edge

**Review by n0izn0iz — CHANGES_REQUESTED** [review]

File: gno/p/dao_core/dao_core_test.gno [hunk 4]
```diff
@@ -73,7 +73,7 @@ func (rm *rolesModule) HasRole(address std.Address, role string) bool {
 	return false
 }
 
-func (rm *rolesModule) NewRole(roleName string) {
+func (rm *rolesModule) NewRole(roleName string, permissionsJSON string) {
```
Comment (MikaelVallenet) [inline gno/p/dao_core/dao_core_test.gno:76]: add a NewRoleJSON func

File: gno/p/dao_roles_voting_group/roles_voting_group_test.gno [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package dao_roles_voting_group
+
+import (
+	"testing"
+
+	dao_interfaces "gno.land/p/teritori/dao_interfaces"
+)
+
```
Comment (MikaelVallenet) [inline gno/p/dao_roles_voting_group/roles_voting_group_test.gno:8]: would be nice to add some specific test about roles voting

File: gno/r/dao_realm/dao_realm.gno [hunk 6]
```diff
@@ -1,5 +1,7 @@
 package dao_realm
 
+// TODO: Create two dao_realm example: Membership based & Roles based
```
Comment (MikaelVallenet) [inline gno/r/dao_realm/dao_realm.gno:3]: Do it

File: packages/screens/Organizations/components/RolesOrg/RolesReviewInformationSection.tsx [hunk 7]
```diff
@@ -169,7 +169,9 @@ export const RolesReviewInformationSection: React.FC<
             <ReviewCollapsableItem
               title={`ROLE #${index + 1}`}
               value={() => (
-                <BrandText style={fontSemibold14}>{role.name}</BrandText>
+                <BrandText style={fontSemibold14}>
+                  {role.name} features: {role.resources?.join(", ")}
```
Comment (MikaelVallenet) [inline packages/screens/Organizations/components/RolesOrg/RolesReviewInformationSection.tsx:173]: put in the desc as smth to enhance in another PR/later

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx [hunk 8]
```diff
@@ -118,3 +216,47 @@ export const RolesSettingsSection: React.FC<RolesSettingsSectionProps> = ({
     </View>
   );
 };
+
+// TODO: Create a hook to get all the resources
+const fakeResources = [
```
Comment (MikaelVallenet) [inline packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx:221]: This is the next PR in coming

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx [hunk 9]
```diff
@@ -26,25 +29,119 @@ interface RolesSettingsSectionProps {
...
+            noBrokenCorners
+            name={`roles.${rolesIndexes.length}.color`}
+            label="Role color"
+            placeholder="Role color"
+            placeHolder="Role color"
+          />
+        </View>
+        <SpacerColumn size={2.5} />
+        <View>
+          <Label>Features & Permissions</Label>
```
Comment (MikaelVallenet) [inline packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx:115]: change the label to Resources

File: packages/utils/gnodao/deploy.ts [hunk 10]
```diff
@@ -19,149 +27,19 @@ interface GnoDAOConfig {
...
   conf: GnoDAOConfig,
 ) => {
-  const source = generateDAORealmSource(networkId, conf);
+  let source = "";
+  if (structure === DaoType.MEMBER_BASED) {
+    source = generateMembershipDAOSource(networkId, conf);
+  } else {
+    source = generateRolesDAOSource(networkId, conf);
+  }
+  console.log(source);
```
Comment (MikaelVallenet) [inline packages/utils/gnodao/deploy.ts:42]: remove this console.log

File: packages/utils/gnodao/generateMembershipDAOSource.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
+import { GnoDAOConfig } from "./deploy";
+import { mustGetGnoNetwork } from "../../networks";
+
+// TODO: Allow the role modules to be optional and don't use in MembershipDAO
```
Comment (MikaelVallenet) [inline packages/utils/gnodao/generateMembershipDAOSource.ts:4]: write in enhancement/open issue but not the scope of this PR

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx [hunk 12]
```diff
@@ -26,25 +29,119 @@ interface RolesSettingsSectionProps {
...
+
+  const onCheckboxChange = (index: number) => {
+    const copyResources = [...resources];
+    copyResources[index].value = !copyResources[index].value;
+    setResources(copyResources);
   };
 
   return (
     <View style={{ flex: 1 }}>
+      <ModalBase
```
[the remark this answers is not in the document]
Comment (WaDadidou) [reply packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx:86]: Only after having the good types to use.
Then, we can make a `NewRoleModal` component with a `NewRoleFormType`
And hugely simplify that: https://github.com/TERITORI/teritori-dapp/blob/1ea963d66838c46bf44fcb5674fc4f42221d5209/packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx#L35-L72

**Review by WaDadidou — CHANGES_REQUESTED** [review]
Partial

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx [hunk 13]
```diff
@@ -26,25 +29,119 @@ interface RolesSettingsSectionProps {
...
+              {/* TODO: Refactor Checkbox to make it a global component instead of Dapp!*/}
+              {resources.map((resource, index) => (
+                <View style={{ flex: 1, flexDirection: "row" }}>
+                  <TouchableOpacity onPress={() => onCheckboxChange(index)}>
+                    <CheckboxDappStore isChecked={resource.value} />
+                  </TouchableOpacity>
+                  <SpacerRow size={1} />
+                  <BrandText style={fontSemibold18}>{resource.name}</BrandText>
+                  <SpacerColumn size={4} />
+                </View>
```
Comment (WaDadidou) [inline packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx:134]: You can `import Checkbox from "@/screens/Mini/components/Checkbox/Checkbox";
`
(Yes, we must export that from `Mini`, it makes non sense)

```suggestion
                <>
                <Checkbox
                  isChecked={resource.value}
                  value={resource.name}
                  label={resource.name}
                  onPress={() => onCheckboxChange(index)}
                />
                  <SpacerColumn size={1} />
                </>
```

**Review by WaDadidou — CHANGES_REQUESTED** [review]

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx [hunk 14]
```diff
@@ -26,25 +29,119 @@ interface RolesSettingsSectionProps {
...
+          <Box
+            style={{
+              borderWidth: 1,
+              borderColor: neutral33,
+              borderRadius: 8,
+              padding: 10,
+            }}
+          >
+            <ScrollView style={{ maxHeight: 180 }}>
+              {/* TODO: Refactor Checkbox to make it a global component instead of Dapp!*/}
```
[the remark this answers is not in the document]
Comment (WaDadidou) [reply packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx:125]: Can be fixed with this change: https://github.com/TERITORI/teritori-dapp/pull/1443#discussion_r1887830135

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx (hunk 13 above)
Comment (MikaelVallenet) [reply packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx:134]: I think i tried the mini but it was smaller and looked weird but i will see

**Comment by MikaelVallenet** [comment]
> Can we open an hackmd for next priorities ? I have some suggestions too (pure UI for the moment) And better to be perfectly aligned 👍

Yes sure i will open this ASAP

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (MikaelVallenet) [reply packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx]: [91b110f](https://github.com/TERITORI/teritori-dapp/pull/1443/commits/91b110f902ad864b6fdbfd13d87376aa09f3d4bc)

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx (hunk 13 above)
Comment (MikaelVallenet) [reply packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx]: [e254883](https://github.com/TERITORI/teritori-dapp/pull/1443/commits/e254883da1271e065957001bdce49470c48d5999)

File: packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (MikaelVallenet) [reply packages/screens/Organizations/components/RolesOrg/RolesSettingsSection.tsx]: [e254883](https://github.com/TERITORI/teritori-dapp/pull/1443/commits/e254883da1271e065957001bdce49470c48d5999)

File: gno/p/dao_core/dao_core.gno [hunk 15]
```diff
@@ -48,16 +48,17 @@ func NewDAOCore(
...
-	core.votingModule = votingModuleFactory(core)
-	if core.votingModule == nil {
-		panic("voting module factory returned nil")
-	}
-
+	// important to keep this order since voting module might depend on roles module
 	core.rolesModule = rolesModuleFactory(core)
 	if core.rolesModule == nil {
 		panic("roles module factory returned nil")
 	}
```
[the remark this answers is not in the document]
Comment (MikaelVallenet) [reply gno/p/dao_core/dao_core.gno:55]: to inform ppl, we decided to do this improvement in a next PR

**Comment by n0izn0iz** [comment]
aussi, sur le fondement de l'article 49 alinéa 3 de la constitution, j'engage la responsabilité de mon organisation décentralisée sur l'application de la requête de fusionnement numéro 1443

**Pull request merged**